File: FirstAndLastOccurence.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> searchRange(vector<int>& nums, int target)
{
    int low = 0, high = nums.size() - 1, mid;
    vector<int> res {-1, -1};

    while (low <= high) {
        mid = (low + (high - low) / 2);
        if ((target == nums[mid]) && ((mid == low) || (target > nums[mid - 1]))) {
            res[0] = mid;
            break;
        }
                
        if (target <= nums[mid]) {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }

    if (res[0] == -1) {
        return res;
    }
        
    low = res[0];
    high = nums.size() - 1;
    while (low <= high) {
        mid = (low + (high - low) / 2);
        if ((target == nums[mid]) && ((mid == high) || (target < nums[mid + 1]))) {
            res[1] = mid;
            break;
        }
                
        if (target >= nums[mid]) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
                
    return res;
}

int main ()
{
    vector<int> nums = {1,3,4,5,5,5,8,9};
    vector<int> res;

    res = searchRange(nums, 5);
    cout << "5 is at [" << res[0] << "," << res[1] << "]" << endl;

    return 0; 
}

File: FirstAndLastOccurence.py
#Time Complexity: O(logn)
#Space Complexity: O(1)
#Successfully ran on leetcode

class Solution:
    def searchRange(self, nums: list(int), target: int) -> list(int):
        if len(nums)==0:
            return [-1,-1]
        first = self.firstOccurence(nums,target)
        last = self.lastOccurence(nums,target)
        return [first,last]
    def firstOccurence(self,nums:list(int),target:int) -> int:
        self.low = 0
        self.high = len(nums)-1
        while self.low<=self.high:
            self.mid = (self.low+self.high)//2
            if target == nums[self.mid]:
                if nums[self.mid-1]!=target or self.mid==self.low:
                    return self.mid
                self.high = self.mid-1
            elif target > nums[self.mid]:
                self.low = self.mid + 1
            else:
                self.high = self.mid-1
        return -1
    def lastOccurence(self,nums:list(int),target:int) -> int:
        self.low = 0
        self.high = len(nums)-1
        while self.low<=self.high:
            self.mid = (self.low+self.high)//2
            if target == nums[self.mid]:
                if self.mid==self.high or nums[self.mid+1]!=target:
                    return self.mid
                self.low = self.mid+1
            elif target > nums[self.mid]:
                self.low = self.mid + 1
            else:
                self.high = self.mid-1
        return -1